src/index.js

```diff
@@ -2,19 +2,38 @@ function defaultEqualityCheck(a, b) {
   return a === b
 }
 
+function areArgumentsShallowlyEqual(equalityCheck, prev, next) {
+  if (prev === next) {
+    return true
+  }
+
+  if (prev === null || next === null || prev.length !== next.length) {
+    return false
+  }
+
+  // Do this in a for loop (and not a `forEach` or an `every`) so we can determine equality as fast as possible.
+  const length = prev.length
+  for (let i = 0; i < length; i++) {
+    if (!equalityCheck(prev[i], next[i])) {
+      return false
+    }
+  }
+
+  return true
+}
+
 export function defaultMemoize(func, equalityCheck = defaultEqualityCheck) {
   let lastArgs = null
   let lastResult = null
-  return (...args) => {
-    if (
-      lastArgs !== null &&
-      lastArgs.length === args.length &&
-      args.every((value, index) => equalityCheck(value, lastArgs[index], index))
-    ) {
-      return lastResult
+
+  // we reference arguments instead of spreading them for performance reasons
+  return function () {
+    if (!areArgumentsShallowlyEqual(equalityCheck, lastArgs, arguments)) {
```

Comment:
The only work-around (that I know of) for referencing arguments in a closure scope is to copy those arguments which is what the original code was doing. Ironically, removing the code that allowed chrome to optimize those functions dramatically sped up the code. I'll whip up a quick example and post it to another branch for you.